# DataFrames

Throughout the course, most of our data exploration will be done with DataFrames. DataFrames are an extremely powerful tool and a natural extension of the Pandas Series. By definition all a DataFrame is:

**A Pandas DataFrame consists of multiple Pandas Series that share index values.**

## Imports

In [5]:
import numpy as np
import pandas as pd

### Creating a DataFrame from Python Objects

In [29]:
# help(pd.DataFrame)

In [30]:
# Make sure the seed is in the same cell as the random call
# https://stackoverflow.com/questions/21494489/what-does-numpy-random-seed0-do
np.random.seed(101)
mydata = np.random.randint(0,101,(4,3))

In [ ]:
mydata

In [32]:
myindex = ['CA','NY','AZ','TX']

In [33]:
mycolumns = ['Jan','Feb','Mar']

In [ ]:
df = pd.DataFrame(data=mydata)
df

In [ ]:
df = pd.DataFrame(data=mydata,index=myindex)
df

In [ ]:
df = pd.DataFrame(data=mydata,index=myindex,columns=mycolumns)
df 

In [ ]:
df.info()


# Reading a .csv file for a DataFrame

----

## NOTE: We will go over all kinds of data inputs and outputs (.html, .csv, .xlxs , etc...) later on in the course! For now we just need to read in a simple .csv file.

----

## CSV
Comma Separated Values files are text files that use commas as field delimeters.<br>
Unless you're running the virtual environment included with the course, you may need to install <tt>xlrd</tt> and <tt>openpyxl</tt>.<br>
In your terminal/command prompt run:

    conda install xlrd
    conda install openpyxl

Then restart Jupyter Notebook.
(or use pip install if you aren't using the Anaconda Distribution)

### Understanding File Paths

You have two options when reading a file with pandas:

1. If your .py file or .ipynb notebook is located in the **exact** same folder location as the .csv file you want to read, simply pass in the file name as a string, for example:
    
        df = pd.read_csv('some_file.csv')
        
2. Pass in the entire file path if you are located in a different directory. The file path must be 100% correct in order for this to work. For example:

        df = pd.read_csv("C:\\Users\\myself\\files\\some_file.csv")

#### Print your current directory file path with pwd

In [ ]:
pwd

#### List the files in your current directory with ls

In [ ]:
ls

In [6]:
df = pd.read_csv('tips.csv')

In [7]:
df

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID  
0    Christy Cunningham  3560325168603410    Sun2959  
1        Douglas Tucker  4478071379779230    Sun4608  
2        Travis Walters  6011812112971322    Sun4458  
3      Nathaniel Harris  4676137647685994    Sun5260  
4          Tonya Carter  4832732618637221    Sun2251  
..                  ...               ...        ...  
239       Michael Avila  5296068606052842    Sat2657  
240      Monica Sanders  3506806155565404    Sat1766  
241          Keith Wong  6011891618747196    Sat3880  
242        Dennis Dixon     4375220550950      Sat17  
243     Michelle Hardin  3511451626698139    Thur672  

[244 rows x 11 columns]

----
About this DataSet (in case you are interested)

* Description
    * One waiter recorded information about each tip he received over a period of a few months working in one restaurant. He collected several variables:

* Format
    * A data frame with 244 rows and 7 variables

* Details
    * tip in dollars,
    * bill in dollars,
    * sex of the bill payer,
    * whether there were smokers in the party,
    * day of the week,
    * time of day,
    * size of the party.

In all he recorded 244 tips. The data was reported in a collection of case studies for business statistics (Bryant & Smith 1995).

* References
    * Bryant, P. G. and Smith, M (1995) Practical Data Analysis: Case Studies in Business Statistics. Homewood, IL: Richard D. Irwin Publishing:
    
* Note: We created some additional columns with Fake data, including Name, CC Number, and Payment ID.

----

# DataFrames

## Obtaining Basic Information About DataFrame

In [8]:
df.columns

Index(['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size',
       'price_per_person', 'Payer Name', 'CC Number', 'Payment ID'],
      dtype='object')

In [ ]:
df.index

In [ ]:
df.head(3)

In [ ]:
df.tail(3)

In [ ]:
df.info()

In [ ]:
len(df)

In [ ]:
df.describe()

In [ ]:
df.describe().transpose()

## Selection and Indexing

Let's learn how to retrieve information from a DataFrame.

### COLUMNS

We will begin be learning how to extract information based on the columns

In [ ]:
df.head()

#### Grab a Single Column

In [ ]:
df['total_bill']

In [ ]:
type(df['total_bill'])

#### Grab Multiple Columns

In [ ]:
# Note how its a python list of column names! Thus the double brackets.
df[['total_bill','tip']]

#### Create New Columns

In [54]:
df['tip_percentage'] = 100* df['tip'] / df['total_bill']

In [ ]:
df.head()

In [56]:
df['price_per_person'] = df['total_bill'] / df['size']

In [ ]:
df.head()

In [ ]:
help(np.round)

#### Adjust Existing Columns

In [59]:
# Because pandas is based on numpy, we get awesome capabilities with numpy's universal functions!
df['price_per_person'] = np.round(df['price_per_person'],2)

In [ ]:
df.head()

#### Remove Columns

In [61]:
# df.drop('tip_percentage',axis=1)

In [62]:
df = df.drop("tip_percentage",axis=1)

In [ ]:
df.head()

# Index Basics

Before going over the same retrieval tasks for rows, let's build some basic understanding of the pandas DataFrame Index.

In [ ]:
df.head()

In [ ]:
df.index

In [ ]:
df.set_index('Payment ID')

In [ ]:
df.head()

In [68]:
df = df.set_index('Payment ID')

In [ ]:
df.head()

In [70]:
df = df.reset_index()

In [ ]:
df.head()

### ROWS

Let's now explore these same concepts but with Rows.

In [ ]:
df.head()

In [73]:
df = df.set_index('Payment ID')

In [ ]:
df.head()

#### Grab a Single Row

In [ ]:
# Integer Based
df.iloc[0]

In [ ]:
# Name Based
df.loc['Sun2959']

#### Grab Multiple Rows

In [ ]:
df.iloc[0:4]

In [ ]:
df.loc[['Sun2959','Sun5260']]

#### Remove Row

Typically are datasets will be large enough that we won't remove rows like this since we won't know thier row location for some specific condition, instead, we drop rows based on conditions such as missing data or column values. The next lecture will cover this in a lot more detail.

In [ ]:
df.head()

In [ ]:
df.drop('Sun2959',axis=0).head()

In [81]:
# Error if you have a named index!
# df.drop(0,axis=0).head()

#### Insert a New Row

Pretty rare to add a single row like this. Usually you use pd.concat() to add many rows at once. You could use the .append() method with a list of pd.Series() objects, but you won't see us do this with realistic real-world data.

In [82]:
one_row = df.iloc[0]

In [ ]:
one_row

In [ ]:
type(one_row)

In [ ]:
df.tail()

In [ ]:
# append is removed from pandas
df.append(one_row).tail()

--------

In [ ]:
df = pd.DataFrame({
    'A': [1, 2, 3],
    'B': [4, 5, 6]
})
new_row = pd.DataFrame({'A': [4], 'B': [7]})
df = pd.concat([df, new_row], ignore_index=True)
df